DFE-Digital/early-careers-framework
# Email redirector

## Purely technical spike.

Currently, in all non-production environments we are sending emails using limited notify API keys that only allows emails ending to a small subset of email addresses. This is absolutely great as we can run the systems without worrying about sending live emails. For testing purposes, we have hardcoded `cpd-test+<alias>` email addresses in the seed data and added those addresses to our team in Notify. There are still a lot of emails that we are not able to see - for example, when testing partnership reporting by lead provider emails are sent to schools email addresses coming from SchoolDataImport which we have no visibility of in PR apps without some heavy DB setup (those emails will silently fail in notify).

This PR introduces an optional redirection for ALL emails sent in non-production environments. Once the environment variable SEND_EMAILS_TO is set, all emails will be sent to given email address instead of the original receiver, which should make QA testing of email journeys much simpler.

Additionally, each email will hold additional tags next to the subject specifying what environment it comes from and to whom it would be sent. However, due to notify limitations (it is not possible to send blank personalisation), this requires specifying subject within a mailer rather than on the template in notify and putting `((subject))` as a subject in the template. I have also tested that all current emails can still be sent (just they will be missing tags in non-production environment)

### Pros:

- All emails are visible during QA with no database and notify group tweaking. For all current and future emails.
- We can redirect emails to our own email address for a local environment - less noise for the team.
- Prevents `Notifications::Client::BadRequestError (BadRequestError: Can’t send to this recipient using a team-only API key)` errors in non-production environments.
- Allows us to turn `raise_notification_errors` in development, which in turn will make it more visible in the development when we are missing 

### Cons:

- Putting subject in the codebase rather than on the template.


## Testing the functionality

1. ssh to the PR box `cf ssh ecf-review-pr-397`
1. `cd ../app && /usr/local/bin/bundle exec rails c`
1. *(optional)* change `ENV['SEND_EMAILS_TO']` to your own `digital.education.gov.uk` address
1. Send a test email: `UserMailer.test_email(user: User.order("RANDOM()").first).deliver_now`
1. Check your box (or cpd-test box if you skipped step 3) - you should see an email with additional tags in the subject

Optionally, try sending any other email - those have not subject personalisation and as such will not have tags when sent, but will still be redirected to given email address.

## Issues to follow

It seems that gmail is still grouping the emails together, despite them having different subjects - causing subject tags to be unreadable/unusable in gmail. This would require more investigating, however good filter should resolve this.
